cli/src/init.rs

```diff
@@ -0,0 +1,238 @@
+use dialoguer::{theme::ColorfulTheme, Input, MultiSelect};
+use include_dir::{include_dir, Dir};
+use std::env;
+use std::error::Error;
+use std::fs::{self, File};
+use std::io::Write;
+use std::path::Path;
+
+pub fn init_project(
+    arg_adapter: &Option<String>,
+    arg_project_name: &Option<String>,
+) -> Result<(), Box<dyn Error>> {
+    let project_name: String = match arg_project_name.as_deref() {
+        None => Input::with_theme(&ColorfulTheme::default())
+            .with_prompt("Project name")
+            .with_initial_text("mopro-example-app".to_string())
+            .interact_text()?,
+        Some(name) => name.to_string(),
+    };
+
+    let adapters = vec!["circom", "halo2"];
+
+    let selection = match arg_adapter.as_deref() {
+        None => MultiSelect::with_theme(&ColorfulTheme::default())
+            .with_prompt("Pick the adapters you want to use (multiple selection with space)")
+            .items(&adapters)
+            .interact()
+            .unwrap(),
+        Some(a) => {
+            let mut selection = vec![];
+            for (i, adapter) in adapters.iter().enumerate() {
+                if a.contains(adapter) {
+                    selection.push(i);
+                }
+            }
+            selection
+        }
+    };
+
+    if selection.is_empty() {
+        println!("\x1b[33mNo adapters selected. Use space to select an adapter.\x1b[0m");
+        init_project(arg_adapter, &Some(project_name))?;
+    } else {
+        let current_dir = env::current_dir()?;
+
+        let project_dir = current_dir.join(&project_name);
+        fs::create_dir(&project_dir)?;
+
+        // Change directory to the project directory
+        env::set_current_dir(&project_dir)?;
+        const TEMPLATE_DIR: Dir = include_dir!("$CARGO_MANIFEST_DIR/src/template/init");
+
+        copy_embedded_dir(&TEMPLATE_DIR, &project_dir, selection.clone())?;
+
+        if let Some(cargo_toml_path) = project_dir.join("Cargo.toml").to_str() {
+            let _ = replace_project_name(cargo_toml_path, &project_name);
+            let _ = replace_features(
+                cargo_toml_path,
+                selection.iter().map(|&i| adapters[i]).collect(),
+            );
+            if selection.contains(&0) {
+                // circom is selected
+                let _ = circom_dependencies_template(cargo_toml_path);
```

Comment:
Why are you assigning these? Why not just drop the value?

```rust
// e.g. just call without assignment
circom_dependencies_template(cargo_toml_path);
```

Reply:
there are warnings here
```sh
unused `Result` that must be used
this `Result` may be an `Err` variant, which should be handled
`#[warn(unused_must_use)]` on by defaultrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Fzhengyawen%2FDocuments%2FGitHub%2Fmopro%2Fcli%2Fsrc%2Finit.rs)
init.rs(63, 17): use `let _ = ...` to ignore the resulting value: `let _ = `
```

but I can use `?` to void this warning as well

Reply:
Ah yes, you want to either use the variable, or use `?`. The current implementation will silence any errors.

In the current implementation if `circom_dependencies_template` returns an error the program will continue as normal. You should use `?` or something like this:

```rust
if let Err(e) = circom_dependencies_template(cargo_toml_path) {
    println!("error in circom dependencies template {e}");
}
```

to unwrap the error

Reply:
It should be caught here
https://github.com/zkmopro/mopro/blob/da913a0ff3b199b33060535f344c473e9e285455/cli/src/main.rs#L88
if there is any error in this file (even with a `?`)

Reply:
No, errors are only propagated if you use the `?` operator, or return the value.

So either of these would work:
```rust
return circom_dependencies_template(cargo_toml_path);
```
```rust
circom_dependencies_template(cargo_toml_path)?;
```

Bot not assigning a variable then not using it. [Here's an example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=511128252eb58b1c1b4265e500abbca9) of the problem.

Reply:
[132638c](https://github.com/zkmopro/mopro/pull/235/commits/132638c70c3e17db08632bd62a1933efe05006b7)